### PR TITLE
Fix PTL2.0 related ASR bugs in r1.21.0: Val metrics logging, None dataloader issue

### PR DESCRIPTION
**Pull request by KunalDhawan**
# What does this PR do ?

This PR added fixes for PTL2.0 related ASR bugs in r1.21.0: Val metrics logging, None dataloader issue

**Collection**: 
ASR, Core

# Changelog 
- val_dataloader() and test_dataloader() functions in nemo/core/classes/modelPT.py updated to handle cases where dataloader is None
- correct logging of validation metrics in nemo/collections/asr/models/rnnt_models.py and nemo/collections/asr/models/hybrid_rnnt_ctc_models.py

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

# Additional Information
* Related to [Bug4296343](https://nvbugswb.nvidia.com/NvBugs5/SWBug.aspx?bugid=4296343&cmtNo=), [Bug4296057](https://nvbugswb.nvidia.com/NvBugs5/SWBug.aspx?bugid=4296057&cmtNo=), [Bug4246804](https://nvbugswb.nvidia.com/NvBugs5/SWBug.aspx?bugid=4246804&cmtNo=)
* Fixes issues raised in #7450, #7507

### CODE REVIEW

File: nemo/collections/asr/models/hybrid_rnnt_ctc_models.py [hunk 1]
```diff
@@ -601,6 +601,9 @@ def validation_pass(self, batch, batch_idx, dataloader_idx):
         if AccessMixin.is_access_enabled():
             AccessMixin.reset_registry(self)
 
+        # adding this as return values are no longer logger automatically in PTL2.0
```
Comment (nithinraok) [inline nemo/collections/asr/models/hybrid_rnnt_ctc_models.py:604]: remove the comments

**Review by nithinraok — COMMENTED** [review]
LGTM.. did you check for other models as well?

File: nemo/collections/asr/models/hybrid_rnnt_ctc_models.py (hunk 1 above)
Comment (KunalDhawan) [reply nemo/collections/asr/models/hybrid_rnnt_ctc_models.py]: Fixed in latest commit. Also added necessary logging changes for label_models, slu_models, ssl_models

Comment (nithinraok) [reply nemo/collections/asr/models/hybrid_rnnt_ctc_models.py]: Thanks Kunal. @athitten how is this different from adding to `test_step_outputs` or `validation_step_outputs` ?

File: nemo/collections/asr/models/ssl_models.py [hunk 2]
```diff
@@ -554,9 +554,11 @@ def validation_step(self, batch, batch_idx, dataloader_idx=0):
...
-            'val_loss': loss_value,
-        }
+        val_log_dict = {'val_loss': loss_value}
+
+        self.log_dict(val_log_dict)
+
+        return val_log_dict
 
     def multi_validation_epoch_end(self, outputs, dataloader_idx: int = 0):
         val_loss_mean = torch.stack([x['val_loss'] for x in outputs]).mean()
```
Comment (stevehuang52) [inline nemo/collections/asr/models/ssl_models.py:564]: Is this one the loss value that should be logged? the one in validation_step is per-batch, but we need to monitor the mean loss over all validation data

File: nemo/collections/asr/models/label_models.py [hunk 3]
```diff
@@ -373,14 +373,17 @@ def evaluation_step(self, batch, batch_idx, dataloader_idx: int = 0, tag: str =
...
-        return {
+        eval_dict = {
             f'{tag}_loss': loss_value,
             f'{tag}_correct_counts': correct_counts,
             f'{tag}_total_counts': total_counts,
             f'{tag}_acc_micro_top_k': acc_top_k,
             f'{tag}_acc_macro_stats': stats,
         }
 
+        self.log_dict(eval_dict)
```
Comment (stevehuang52) [inline nemo/collections/asr/models/label_models.py:384]: not all variables in `eval_dict` need logging, please move `self.log()` into `multi_evaluation_epoch_end` where it calculates the averaged metrics

File: nemo/collections/asr/models/ssl_models.py (hunk 2 above)
Comment (KunalDhawan) [reply nemo/collections/asr/models/ssl_models.py:564]: The logging has to be done at validation step itself akin to the change introduced in this PR for PTL upgrade - https://github.com/NVIDIA/NeMo/pull/6433/files#diff-b2780d88910b132d177fb0081453ad276c5e4aefe47a87f219e96f38af0625be

Comment (KunalDhawan) [reply nemo/collections/asr/models/ssl_models.py:564]: If we do the logging at multi_validation_epoch_end and multi_test_epoch_end, we still get the current error - `ModelCheckpoint(monitor='val_wer')` could not find the monitored key in the returned metrics: ['train_loss', 'learning_rate', 'global_step', 'train_backward_timing in s', 'train_step_timing in s', 'training_batch_wer', 'epoch', 'step']. HINT: Did you call `log('val_wer', value)` in the `LightningModule`?

Comment (KunalDhawan) [reply nemo/collections/asr/models/ssl_models.py:564]: I'm refactoring the PR currently to make the logging similar to how we do it for ctc_models. I'll make the change for RNNT and Hybrid models for now, maybe we can open another PR next to address these issues for the SLU, SSL and label models

File: nemo/collections/asr/models/label_models.py (hunk 3 above)
Comment (KunalDhawan) [reply nemo/collections/asr/models/label_models.py:384]: I think all these variables were logged previously too (example run - https://wandb.ai/nvidia/titanet-chime7-training?workspace=user-kdhawan), please let me know if you want me to remove some of these from the log

File: nemo/collections/asr/models/ssl_models.py (hunk 2 above)
Comment (KunalDhawan) [reply nemo/collections/asr/models/ssl_models.py:564]: New PR - https://github.com/NVIDIA/NeMo/pull/7531

**Comment by KunalDhawan** [comment]
Opened a new PR for this issue - https://github.com/NVIDIA/NeMo/pull/7531
